### PR TITLE
[DO NOT MERGE] Adjust Logitech keyboard lighting when starting a game

### PR DESCRIPTION
**Pull request by BhaaLseN**
Logitech got a range of gaming keyboards that have full-RGB per-key lighting, which allows for some interresting configurations (ex. lighting only the currently configured keys, and assigning them colors per role).
Heres a brief and shaky video showing this in action: https://youtu.be/ZbyJRzliss0

As it is now, it will use the following configuration (with unused keys being blacked out/unlit):
- D-Pad in gray
- C-Stick in yellow
- A in green
- B in red
- Z in purple
- Main Stick in blue
- Triggers in brown/orange
- everything else thats configured in white

Obviously, it requires the appropriate hardware (such as the G910 "Orion Spark", which I've gotten my hands on lately) and the companion app Logitech Gaming Software (short LGS) with Version 8.55 or higher.
Machines without the appropriate hardware or without LGS will simply ignore any calls and do nothing instead.

Some things to note:
- It's 1p only - there is a TODO in there to support more than one player on the same keyboard, but I might aswell drop that comment...its pretty unlikely that two players would share the same keyboard.
- It doesn't support key combination (such as Ctrl+W), since it would need to be a lot more involved to work. And I guess unless someone requests it, we can put it off until then :P
- I don't intend to provide some sort of configuration interface for it. It's too much work for just picking colors; and LGS has a global configuration switch to turn this on or off. I simply hope that my default choices are good enough for now.
- It won't support non-Logitech devices (obviously, since its the Logitech SDK), and it will be limited to per-key RGB devices (since it doesn't make much sense to change color on all keys at once). There seems to be [a project out there](http://ledl.io/) that also supports Corsair keyboards, but they're far from ready (and not open source yet).
- The SDK is Windows only, with no plans to support Linux or OSX in the near term. I didn't find any viable solutions for non-Windows either, so this will be a Windows only feature for the time being. ```g15daemon``` has some custom code that throws USB commands and magic numbers around, but nothing I could adopt (let alone, read and understand) to work with this. Those OS will use ```LEDStub.cpp``` which does nothing.

As for the review part:
- I couldn't find any more appropriate place to hook my initialization code in (I would've preferred one that is called on all OS and devices), so I had to put some in both ```Frame.cpp``` and ```MainNoGUI.cpp```.
- Same applies to starting and stopping the effect; which might not be in the best place there in ```Core.cpp``` (yet again, I couldn't find a surefire place that worked 100% other than that).
- I couldn't find a more elegant solution for initialization of the key mappings and assigning colors, feel free to suggest something better that doesn't duplicate values.
- And if you know any better, please suggest some modifiers (or even different containers) for the static members in ```LEDWin.cpp```. Making the maps ```const``` had some odd compiler problems, so I had to drop them.

### CODE REVIEW

File: Externals/Logitech/Include/LogitechLEDLib.h [hunk 1]
```diff
@@ -0,0 +1,171 @@
+//
+// Logitech Gaming LED SDK
+//
+// Copyright (C) 2011-2014 Logitech. All rights reserved.
```
Comment (delroth) [inline Externals/Logitech/Include/LogitechLEDLib.h]: License?

File: Source/Core/Core/LEDWin.cpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	// due to their limited possibilities of displaying control state.
+	LogiLedSetTargetDevice(LOGI_DEVICETYPE_PERKEY_RGB);
+	// black out all keys initially
+	LogiLedSetLighting(0, 0, 0);
+
+	InputConfig* const pad_plugin = Pad::GetConfig();
+	// grab the Player 1 controller (TODO: what about other local players?)
+	ControllerEmu* const controller = pad_plugin->controllers[0];
+
+#define COLOR_PERCENT(c) ((c).red * 100 / 0xFF), ((c).green * 100 / 0xFF), ((c).blue * 100 / 0xFF)
```
Comment (delroth) [inline Source/Core/Core/LEDWin.cpp]: Why this #define? You only use it once.

File: Externals/Logitech/Include/LogitechLEDLib.h (hunk 1 above)
Comment (BhaaLseN) [inline Externals/Logitech/Include/LogitechLEDLib.h]: Free to use, I specifically mailed their support and that guy replied "Please feel free to include our header file and the static library in your project. There's already plenty of publicly available repositories that share our SDK."
Nothing more specific though.

File: Source/Core/Core/LEDWin.cpp (hunk 2 above)
Comment (BhaaLseN) [inline Source/Core/Core/LEDWin.cpp]: Mostly for being too lazy to type the ```s_colors[idx].col``` part all the time. Had multiple uses initially, can drop it.

File: Externals/Logitech/Include/LogitechLEDLib.h (hunk 1 above)
Comment (Tilka) [inline Externals/Logitech/Include/LogitechLEDLib.h]: Some GPL-incompatible EULA.

Comment (BhaaLseN) [inline Externals/Logitech/Include/LogitechLEDLib.h]: That'd almost be a shame. Which ones in particular?

File: Source/Core/Core/LEDWin.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+	// 7: brown (L/R Triggers, to distinguish them from other keys)
+	{ 0x80, 0x35, 0x00 },
+};
+
+// overrides for specific named keys.
+// anything not in this list will use s_colors[0] as default color.
+// key is derived from a combination of
+// ControllerEmu::ControlGroup::name and ControllerEmu::ControlGroup::Control::name.
+// an example would be "Buttons/A" or "C-Stick/Up"
+static std::map<std::string, int> s_named_key_colors =
```
Comment (lioncash) [inline Source/Core/Core/LEDWin.cpp]: These maps should be const. You can use the `.at` method to access elements.

**Comment by lioncash** [comment]
License issues aside, while I think this is a neat idea, I don't particularly agree with the implementation. This isn't very extensible, and introduces non-emulation-related code into Core – something that would ideally be something that should be abstracted away in InputCommon.

File: Externals/Logitech/Include/LogitechLEDLib.h (hunk 1 above)
Comment (phire) [inline Externals/Logitech/Include/LogitechLEDLib.h]: GPL conflicts with the "nontransferable" part of that EULA. It also requires the right to modify the SDK (and have source provided to do so)

Still, better than the sdk to integrate valve's steam controller.

File: Source/Core/Core/LEDWin.cpp (hunk 3 above)
Comment (BhaaLseN) [inline Source/Core/Core/LEDWin.cpp]: As I mentioned, VS threw some obscure errors (like ```std::map<foo> has no matching operator[]```) when I tried that, but that might have been something else.

**Comment by BhaaLseN** [comment]
Closing for now, at least until the licensing issues are addressed. The branch will stay up on my remote tho (for reference).

Regarding the dependency on Core, I wish I could have found a better way of doing this, but for the most part it either came too late into the initialization or not at all, leaving me with randomly colored keys. I'll meditate a bit and see what I can do about it.

**Pull request closed without merging**